### ISSUE TITLE
[clang-tidy] cppcoreguidelines-pro-type-member-init check flags anonymous union member

### ISSUE THREAD

**Issue by paulaltin** [llvm/llvm-project#54748]
Running clang-tidy on this code:

```
struct s {
    union {
        int a = 0;
        int b;
    };
};
```

results in the following warning:

```
warning: constructor does not initialize these fields: b [cppcoreguidelines-pro-type-member-init]
```

However, adding an initializer to `b` (e.g. `int b = 0`) results in a clang diagnostic:

```
error: initializing multiple members of union [clang-diagnostic-error]
```

**Comment by paulaltin** [comment]
The cppcoreguidelines-pro-type-member-init.cpp file has a test that ensures the check does not flag a union with only one member initialized:

https://github.com/llvm/llvm-project/blob/6f73bd781305266a747055875ce8352e5a36c809/clang-tools-extra/test/clang-tidy/checkers/cppcoreguidelines-pro-type-member-init.cpp#L296

and a test that ensures the check does flag members of anonymous unions when none are initialized:

https://github.com/llvm/llvm-project/blob/6f73bd781305266a747055875ce8352e5a36c809/clang-tools-extra/test/clang-tidy/checkers/cppcoreguidelines-pro-type-member-init.cpp#L332

(both authored by @alexfh).

**Comment by paulaltin** [comment]
A similar issue was reported in 2017: https://github.com/llvm/llvm-project/issues/34232

At the time, @JonasToth suggested the bug might be related to this commit: https://github.com/llvm-mirror/clang-tools-extra/commit/b7c157e0b8f327221bd81cf0694bdbe76352f766

Since then, a commit by @Sockke was added to fix an apparently related issue: https://github.com/llvm/llvm-project/commit/e4902480f1e2f12f73c2b504e3d717536653dd7b

**Comment by dfrib** [comment]
The same warning is emitted also for when initializing one of the union members in a mem-init-list (as opposed to default mem initializers in the union), flagging e.g. the idiomatic std::optional backports for constexpr default-construction of disengaged optionals:
```cpp
struct Optional {
    // warning: constructor does not initialize these fields: value_ 
    // [cppcoreguidelines-pro-type-member-init,hicpp-member-init]
    constexpr Optional() : dummy_{} {}
    union {
        char dummy_;
        int value_;
    };
};
```

**Comment by movie-travel-code** [comment]
Btw, the review differential of the fix patched by @Sockke is https://reviews.llvm.org/D108370. IMHO, clang-tidy should be conservative when emitting the diagnostics, if the clang-tidy cannot afford the overhead or has no confidence to cover all the cases,  it should give up emitting the warnings.

**Comment by Sockke** [comment]
The latest patch is here, and there are still omissions in the processing of a union.

**Comment by Sockke** [comment]
Fixed

**Comment by paulaltin** [comment]
Thanks for your work fixing this and getting it through review @Sockke!
